src/JustEat.StatsD/PooledUdpTransport.cs

```diff
@@ -40,66 +99,58 @@ public void Send(string metric)
 
             var bytes = Encoding.UTF8.GetBytes(metric);
             var endpoint = _endpointSource.GetEndpoint();
+            var pool = GetPool(endpoint);
+            var socket = pool.Pool.PopOrCreate();
 
-            var socket = _socketPool.PopOrCreate();
             try
             {
-                socket.SendTo(bytes, endpoint);
+                socket.Send(bytes);
             }
             catch (Exception)
             {
                 socket.Dispose();
                 throw;
             }
 
-            _socketPool.Push(socket);
+            pool.Pool.Push(socket);
         }
-
-        /// <summary>
-        /// Finalizes an instance of the <see cref="PooledUdpTransport"/> class.
-        /// </summary>
-        ~PooledUdpTransport()
-        {
-            Dispose(false);
-        }
-
+        
         /// <inheritdoc />
         public void Dispose()
         {
-            Dispose(true);
-            GC.SuppressFinalize(this);
+            _pool?.Dispose();
         }
 
-        /// <summary>
-        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
-        /// </summary>
-        /// <param name="disposing">
-        /// <see langword="true" /> to release both managed and unmanaged resources;
-        /// <see langword="false" /> to release only unmanaged resources.
-        /// </param>
-        private void Dispose(bool disposing)
+        private readonly object _lock = new object();
+
+        private ConnectedPool GetPool(IPEndPoint endPoint)
         {
-            if (!_disposed)
+            if (_pool == null)
             {
-                try
+                lock (_lock)
                 {
-                    while (_socketPool.Count > 0)
+                    if (_pool == null)
                     {
-                        var socket = _socketPool.Pop();
-                        socket?.Dispose();
+                        _pool = new ConnectedPool(endPoint);
+                        return _pool;
                     }
                 }
-                catch (ObjectDisposedException)
+            }
+
+            if (endPoint.Equals(_pool.IpEndPoint))
+            {
+                return _pool;
+            }
+            else
+            {
+                lock (_lock)
                 {
-                    // If the pool has already been disposed by the finalizer, ignore the exception
-                    if (disposing)
-                    {
-                        throw;
-                    }
+                    _pool.Dispose();
+                    _pool = new ConnectedPool(endPoint);
```

Comment:
If any of the connections in the constructor throw, then subsequent calls will return the disposed-of pool, putting everything in a bad state. Something like an optimistic `Interlocked.CompareExchange()` might be better, as that removes the need for a lock and does not corrupt state if the initialization fails.

Note this doesn't check whether the pool that was created before the thread created its pool is connected to the same endpoint.

```csharp
var oldPool = _pool;
var newPool = new ConnectedPool(endPoint);

try
{
    if (Interlocked.CompareExchange(ref _pool, oldPool, newPool) == oldPool)
    {
        We successfully exchanged the pools
        oldPool.Dispose();
    }
    else
    {
        // Something already replaced the pool
        newPool.Dispose();
    }
}
catch (Exception)
{
    newPool.Dispose();
}
```